File: vdslib/src/vespa/vdslib/state/state.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include <vespa/vdslib/state/state.h>
#include <vespa/vespalib/util/exceptions.h>
#include <ostream>

namespace storage::lib {

const State&
State::get(std::string_view serialized)
{
    if (serialized.size() == 1) switch(serialized[0]) {
        case '-': return UNKNOWN;
        case 'm': return MAINTENANCE;
        case 'd': return DOWN;
        case 's': return STOPPING;
        case 'i': return INITIALIZING;
        case 'r': return RETIRED;
        case 'u': return UP;
        default: break;
    }
    throw vespalib::IllegalArgumentException(
            "Unknown state " + serialized + " given.", VESPA_STRLOC);
}

State::State(std::string_view name, std::string_view serialized,
             uint8_t rank,
             bool validDistributorReported, bool validStorageReported,
             bool validDistributorWanted, bool validStorageWanted,
             bool validCluster)
    : _name(name),
      _serialized(serialized),
      _rankValue(rank),
      _validReportedNodeState(2),
      _validWantedNodeState(2),
      _validClusterState(validCluster)
{
        // Since this is static initialization and NodeType is not necessarily
        // created yet, use these enum values.
    uint32_t storage = 0;
    uint32_t distributor = 1;
    _validReportedNodeState[storage] = validStorageReported;
    _validReportedNodeState[distributor] = validDistributorReported;
    _validWantedNodeState[storage] = validStorageWanted;
    _validWantedNodeState[distributor] = validDistributorWanted;
}

State::~State() = default;

void
State::print(std::ostream& out, bool verbose, const std::string& indent) const
{
    (void) indent;
    out << (verbose ? _name : _serialized);
}

}

File: vdslib/src/main/java/com/yahoo/vdslib/state/State.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.vdslib.state;

import java.util.ArrayList;

/**
 *
 * Defines legal states for various uses. Split this into its own class such
 * that we can easily see what states are legal to use in what situations.
 * They double as disk states and node states nodes report they are in, and
 * wanted states set external sources.
 */
public enum State {

    // The order declares the ordinals, and defines what states are above/below others
    UNKNOWN     ("-", true, true, false, false, false, false, false), // This state is used by the fleetcontroller to indicate
                                                                      // that we have failed to contact the node. It should never be
                                                                      // sent out of the fleetcontroller
    MAINTENANCE ("m", false, false, true, true, false, true, true),
    DOWN        ("d", true, true, true, true, true, true, true), // Down is not valid reported state sent from the node itself.
    STOPPING    ("s", true, true, false, false, true, true, true),
    INITIALIZING("i", true, true, false, false, true, true, true),
    RETIRED     ("r", false, false, false, true, false, true, true),
    UP          ("u", true, true, true, true, true, true, true);

    private final boolean validClusterState;
    private final ArrayList<Boolean> validReportedNodeState = new ArrayList<>();
    private final ArrayList<Boolean> validWantedNodeState = new ArrayList<>();
    private final ArrayList<Boolean> validCurrentNodeState = new ArrayList<>();
    private final String serializedAs;

    State(String serialized, boolean validDistReported, boolean validStorReported, boolean validDistWanted,
          boolean validStorWanted, boolean validCluster, boolean validDistCurrent, boolean validStorCurrent) {
        validClusterState = validCluster;
        validReportedNodeState.add(validStorReported);
        validReportedNodeState.add(validDistReported);
        validWantedNodeState.add(validStorWanted);
        validWantedNodeState.add(validDistWanted);
        validCurrentNodeState.add(validStorCurrent);
        validCurrentNodeState.add(validDistCurrent);
        this.serializedAs = serialized;
    }

    public static State get(String serialized) {
        for (State s : values()) {
            if (s.serializedAs.equals(serialized)) { return s; }
        }
        throw new IllegalArgumentException("Invalid state '" + serialized + "'.");
    }

    public String serialize() { return serializedAs; }

    public boolean validClusterState() { return validClusterState; }
    public boolean validReportedNodeState(NodeType type) { return validReportedNodeState.get(type.ordinal()); }
    public boolean validWantedNodeState(NodeType type) { return validWantedNodeState.get(type.ordinal()); }
    public boolean validCurrentNodeState(NodeType type) { return validCurrentNodeState.get(type.ordinal()); }

    public boolean maySetWantedStateForThisNodeState(State s) { return (s.ordinal() <= ordinal()); }

    public boolean oneOf(String states) {
        for (char c : states.toCharArray()) {
            String s = "" + c;
            if (s.equals(serializedAs)) return true;
        }
        return false;
    }

    @Override
    public String toString() {
        String id = name();
        String lower = id.substring(1).toLowerCase();
        return id.charAt(0) + lower;
    }

}
